Custom shaders

Spongman opened the pull request:
i was playing around and discovered that it's impossible to use a custom shader that doesn't make use of all the vertex attributes that are used in the built-in shaders. this PR loosens this restriction, and adds an example using such a custom shader to the p5.Shader constructor docs.

kjhollen commented:
yay! yes, this was on my list of shortcomings for the current shader implementation—thank you for taking this on. I've messed up my notifications for this project somehow and am sorry I didn't see this sooner—I'll try to review soon!

Spongman commented:
thanks! i just learned that you can add '?w=1' to the end of a diff's URL in github and it'll ignore whitespace. might be helpful in this case...

Spongman commented:
@kjhollen did you ever manage to review this? I have a few other webgl-related PRs that might help close some issues.

Spongman commented:
fixes #2435

Spongman commented:
also fixes #2262

Spongman commented:
Mostly it removes the assumption that all those buffers are used by the shader. Without this fix you get an exception thrown if you try to use a shader that doesn't define all those attributes (eg, try to run the sample in the 2nd commit).

This problem also manifests in some browsers where the attributes are declared in the shader, but are optimized away by the shader compiler. This is what is causing the issues on Firefox/Linux and chrome/android.

I'll take a look at the retained-mode code...

kjhollen commented on src/webgl/p5.RendererGL.Immediate.js:
```diff
@@ -161,25 +161,31 @@ p5.RendererGL.prototype._drawFillImmediateMode = function(
+    );
+  }
+
+  if (
```
@lmccart any style guide suggestions here for multi-line if statement? I'm not sure we've used this convention elsewhere.

Spongman replied:
#2336 added auto-formatting for the whole library. this is how it's configured, apparently.

kjhollen replied:
ok, we will go with what the auto-formatter wants :)

kjhollen commented on src/webgl/p5.RendererGL.Immediate.js:
```diff
@@ -197,7 +203,11 @@ p5.RendererGL.prototype._drawFillImmediateMode = function(
+  if (
+    this.drawMode === constants.TEXTURE &&
+    this.curFillShader.attributes.aTexCoord
+  ) {
```
hmm, if the drawMode is TEXTURE but there are no texture coordinates available, this feels like maybe an error worth noting to the user? (under what circumstances is this edge case possible—is this true if we're drawing lines?)

Spongman replied:
i don't think that's an error. you could certainly have a shader that uses a texture, but does not use the aTexCoord vertex attributes.

kjhollen commented on src/webgl/p5.RendererGL.Immediate.js:
```diff
@@ -221,20 +231,14 @@ p5.RendererGL.prototype._drawFillImmediateMode = function(
   if (this.drawMode === constants.FILL || this.drawMode === constants.TEXTURE) {
     switch (this.immediateMode.shapeMode) {
       case constants.LINE_STRIP:
```
@mlarghydracept is the math for calculating the shapes done elsewhere (are they computed to be triangle fans always? or should we mark this part of the code as TODO for future implementation / given issue #2344 ?)

Spongman replied:
I'm not sure about that. i don't think i changed any logic here...

kjhollen replied:
Right. Something about not having the individual outcomes explicitly (even though they were all the same) made me wonder if this was done on purpose originally as a stop-gap and maybe doesn't indicate that as clearly, updated in this way. It's functionally the same, though!

mlarghydracept replied:
Hey sorry - my notifications are whacky so I sometimes miss my mentions. No changes are made here. The reason things are happening like they do over in #2344 isn't because of gl modes because gl isn't actually rendering those lines. So, counter-intuitively, the constants.LINE_STRIP has nothing to do with that problem.

kjhollen commented on src/webgl/p5.RendererGL.Retained.js:
```diff
@@ -43,127 +56,24 @@ p5.RendererGL.prototype.createBuffers = function(gId, obj) {
-  );
-  //}
-};
+  if (this.curStrokeShader.attributes.aPosition) {
```
this looks great! it occurs to me that now it's probably worth adding a comment about what each if-block is doing here to make the code a little more scannable. we probably should have had this already. :)

kjhollen commented on src/webgl/p5.Shader.js:
```diff
@@ -16,6 +16,47 @@ var p5 = require('../core/core');
+ *   '    z = vec2(z.x*z.x-z.y*z.y, 2.0*z.x*z.y) + c;' +
+ *   '  }' +
+ *   '  gl_FragColor = vec4(0.5-cos(n*17.0)/2.0,0.5-cos(n*13.0)/2.0,0.5-cos(n*23.0)/2.0,1.0);' +
+ *   '}';
```
example needs a short description for accessibility

Spongman replied:
add a comment here is easier said than done ;-)

Spongman replied:
oh, i see you meant an @alt tag, i'll add one...

kjhollen commented on src/webgl/p5.Shader.js:
```diff
@@ -16,6 +16,47 @@ var p5 = require('../core/core');
+ * var mandel;
+ * function setup() {
+ *   var renderer = createCanvas(100, 100, WEBGL);
+ *   mandel = new p5.Shader(renderer, vs, fs);
```
hmm, other docs say that the return type from createCanvas is an HTMLCanvasElement, not an instance of p5.RendererGL?

Spongman replied:
yup. looks like the docs are wrong ;-)

Spongman replied:
maybe it might be nice to add a `createShader()` method that wraps this functionality?

kjhollen replied:
oops :)

there's `loadShader()` but it takes file names! https://p5js.org/reference/#/p5/loadShader, maybe this can be overridden to do what you describe?

documentation for this method needs a good bit of work, but there are some working examples in the manual tests.

Spongman commented:
i added a `createShader()` method that takes the shader sources as strings. and i moved the mandelbrot sample to that and removed the now-unnecessary fetching of the renderer.

i also added some comments to the sample.

kjhollen commented on src/webgl/p5.Shader.js:
```diff
@@ -44,38 +77,28 @@ var p5 = require('../core/core');
-  this.samplers = [];
-
-  return this;
+p5.prototype.createShader = function(vs, fs) {
```
this looks great. the rest of this file only defines the Shader class, but no p5.prototype methods. I think it feels ok here but wonder if it belongs in material.js alongside loadShader? this file has more of the p5 public methods for working with shaders.

kjhollen commented:
just built this on my machine and everything seems great! automated tests are passing, manual tests work (actually one is broken, but is broken in the mainline, too) :) can we just get the createShader method moved over to material.js? then we'll be ready to go. lots of great work in here—THANKS!!!

kjhollen commented:
awesome, THANK YOU!! 🎉

saitogroup commented:
I looked at another thread and it said that error was solved here... but I couldn't see/understand that. I think it has something to do with when you're texturing an object, rather than texturing a primitive, but I'm not entirely sure (some kind of reference error).
